### PR TITLE
Bug 1210131 - Fixes Wrong Favicon shown after tapping back quickly.

### PR DESCRIPTION
**Pull request by farhanpatel**
When tapping back quickly the favicon fails to update because the page is loaded from the WKWebView cache instead of rendered again. This causes the script responsible for fetching the favicons never to be run. Now instead wait till loading is finished and then run the script to fetch favicons.

Sidenote: this also fixes [Bug-1252888](https://bugzilla.mozilla.org/show_bug.cgi?id=1252888) but I'll update that with a separate PR.

### CODE REVIEW

File: Client/Assets/Favicons.js [hunk 1]
```diff
@@ -2,47 +2,47 @@
...
-      for (var i = 0; i < icons.length; i++) {
-        var href = icons[i].href;
-        res[href] = this.selectors[selector];
-          foundIcons = true
+window._FireFox_Favicons = function() {
+    // These integers should be kept in sync with the IconType raw-values
+    var ICON = 0;
+    var APPLE = 1;
+    var APPLE_PRECOMPOSED = 2;
+    var GUESS = 3;
```
Comment (thebnich) [inline Client/Assets/Favicons.js]: Nit: Please keep 2-space indentation for JS files. Sorry -- it's confusing! If you aren't already, I suggest using a separate text editor for JS files other than Xcode.

File: Client/Assets/Favicons.js [hunk 2]
```diff
@@ -2,47 +2,47 @@
...
-    for (selector in this.selectors) {
-      var icons = document.querySelectorAll(selector)
-      if (icons.length) {
-        res = {}
-      }
-      for (var i = 0; i < icons.length; i++) {
-        var href = icons[i].href;
-        res[href] = this.selectors[selector];
-          foundIcons = true
+window._FireFox_Favicons = function() {
```
Comment (thebnich) [inline Client/Assets/Favicons.js]: Please use the `__firefox__` namespace that we use in other content scripts. The initialization pattern we've been using is this:

```
if (!window.__firefox__) {
  window.__firefox__ = {};
}

window.__firefox__.favicons = ...
```

This keeps everything namespaced to a single global object.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 3]
```diff
@@ -1766,6 +1771,7 @@ extension BrowserViewController: TabManagerDelegate {
 
 extension BrowserViewController: WKNavigationDelegate {
     func webView(webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
+
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: Nit: revert this change

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 4]
```diff
@@ -869,6 +869,7 @@ class BrowserViewController: UIViewController {
             guard let loading = change?[NSKeyValueChangeNewKey] as? Bool else { break }
             toolbar?.updateReloadStatus(loading)
             urlBar.updateReloadStatus(loading)
+            if (!loading) {runScriptsOnWebView(webView)}
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: Nit: to follow our formatting conventions, please put the `runScriptsOnWebView` on a new line (like we do for the `log.debug` right below here).

File: Client/Assets/Favicons.js [hunk 5]
```diff
@@ -2,47 +2,49 @@
...
+  var selectors = { "link[rel~='icon']": ICON,
+    "link[rel='apple-touch-icon']": APPLE,
+    "link[rel='apple-touch-icon-precomposed']": APPLE_PRECOMPOSED
+  };
+  
+  function getAll() {
+    var favicons = {};
+    
+    for (var selector in selectors) {
+      var icons = document.head.querySelectorAll(selector);
```
Comment (thebnich) [inline Client/Assets/Favicons.js]: Didn't even notice some of these fixes you added before. Nice :+1:

**Comment by thebnich** [comment]
And also thank you for the quick turnaround for addressing the comments!

**Comment by farhanpatel** [comment]
No problem. Just happened to be at the computer as you were commenting :smile:
